internal/libtor/enabled.go

```diff
@@ -0,0 +1,213 @@
+		err = <-p.startErr
+	})
+	return err
+}
+
+// Wait implements [process.Process].
+func (p *torProcess) Wait() (err error) {
+	p.waitOnce.Do(func() {
+		err = <-p.waitErr
+	})
+	return
+}
+
+var (
+	// ErrTooManyArguments indicates that p.args contains too many arguments
+	ErrTooManyArguments = errors.New("libtor: too many arguments")
+
+	// ErrCannotCreateControlSocket indicates that we cannot create a control socket.
+	ErrCannotCreateControlSocket = errors.New("libtor: cannot create a control socket")
+)
+
+// runtor runs tor until completion and ensures that tor exits when
+// the given ctx is cancelled or its deadline expires.
+func (p *torProcess) runtor(ctx context.Context, cc net.Conn, args ...string) {
+	// wait for Start or context to expire
+	select {
+	case <-p.awaitStart:
+	case <-ctx.Done():
+		return
+	}
+
+	// TODO(bassosimone): I don't understand whether I should decorate these
+	// C pointers using unsafe.Pointer or it doesn't matter
+
+	// Create argc and argv for tor
+	argv := append([]string{"tor"}, args...)
+	const toomany = 256 // arbitrary low limit to make C.int and C.size_t casts always work
+	if len(argv) > toomany {
+		p.startErr <- ErrTooManyArguments // nonblocking channel
+		return
+	}
+	argc := C.size_t(len(argv))
+	cargv := C.cstringArrayNew(argc)
+	defer C.cstringArrayFree(cargv, C.size_t(argc))
+	for idx, entry := range argv {
+		C.cstringArraySet(cargv, C.size_t(idx), C.CString(entry))
+	}
+
+	// Add to config a WEAK REFERENCE to argc and argv
+	config := C.tor_main_configuration_new()
+	runtimex.PanicIfNil(config, "C.tor_main_configuration_new failed")
+	defer C.tor_main_configuration_free(config)
+	code := C.tor_main_configuration_set_command_line(config, C.int(argc), cargv)
+	runtimex.Assert(code == 0, "C.tor_main_configuration_set_command_line failed")
+
+	// Create OWNING file descriptor
+	filedesc := C.tor_main_configuration_setup_control_socket(config)
+	if !C.filedescIsGood(filedesc) {
+		p.startErr <- ErrCannotCreateControlSocket // nonblocking channel
+		return
+	}
+
+	// Convert the OWNING file descriptor into a proper file.
+	filep, err := net.FileConn(os.NewFile(uintptr(filedesc), ""))
+	if err != nil {
+		p.startErr <- err // nonblocking channel
+		return
+	}
+
+	// Make sure we close filep when the context is done. Because the
+	// socket is OWNING, this will also cause tor to return.
+	go func() {
+		defer filep.Close()
+		defer cc.Close()
+		<-ctx.Done()
+	}()
+
+	// TODO(bassosimone): what should we do if the user closes the
+	// control connection? What happens when tor is a subprocess and
+	// we close the control connection? We should do the same!
+
+	// Route messages from and to the control connection
+	go sendrecv(cc, filep)
+	go sendrecv(filep, cc)
+
+	// Tell user that startup is successful.
+	p.startErr <- nil // nonblocking chan
+
+	// Run tor until completion. Note that return codes are not
+	// currently documented and they're never zero (WTF?!).
+	_ = C.tor_run_main(config)
```

Comment:
This is actually wrong: there is a case that I missed (🤦) where the code returns `0`. So we should consider any value different from `0` here to be an execution error for `tor`.